### PR TITLE
Make certificate generation FIPS compliant

### PR DESCRIPTION
**Pull request by weinimo**
This uses PKCS#8 private key encryption now, which is FIPS compliant.

[OSPRH-9709](https://issues.redhat.com//browse/OSPRH-9709)
[OSPRH-6237](https://issues.redhat.com//browse/OSPRH-6237)

### CODE REVIEW

File: pkg/octavia/pem_decrypt_fips.go [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+	keySize:    32,
+	blockSize:  aes.BlockSize,
+},
+}
+
+// deriveKey uses a key derivation function to stretch the password into a key
+// with the number of bits our cipher requires. This algorithm was derived from
+// the OpenSSL source.
+func (c rfc1423Algo) deriveKey(password, salt []byte) []byte {
+	hash := sha256.New()
```
Comment (weinimo) [inline pkg/octavia/pem_decrypt_fips.go:83]: This is the only code line that I have modified.

File: pkg/octavia/pem_decrypt_fips.go [hunk 2]
```diff
@@ -0,0 +1,152 @@
+// Copyright 2012 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
```
Comment (johnsom) [inline pkg/octavia/pem_decrypt_fips.go:3]: Is the license file for the operator the same as the license in the GOlang?
If not, you probably need to paste the golang license here.

File: pkg/octavia/amphora_certs.go [hunk 3]
```diff
@@ -60,12 +60,12 @@ func generateKey(passphrase []byte) (*rsa.PrivateKey, []byte, error) {
 
 	var pemBlock *pem.Block
 	if passphrase != nil {
-		pemBlock, err = x509.EncryptPEMBlock( //nolint:staticcheck
+		pemBlock, err = EncryptPEMBlock(
 			rand.Reader,
 			"PRIVATE KEY",
```
Comment (johnsom) [inline pkg/octavia/amphora_certs.go:65]: This might need to be "ENCRYPTED PRIVATE KEY"

Comment (weinimo) [reply pkg/octavia/amphora_certs.go:65]: Disclaimer: I am by no means an expert in this field, even though I did a significant amount of research while working on this, and the different formats can be confusing, so I might be wrong. However this is my understanding:

The way x509.EncryptPEMBlock works is that it seems to encrypt the PEM encoded data before storing it in the PEM file. IDK how to call this format, but is definitely an outdated way and not what is commonly referred to as PKCS#8. It also is not PKCS#1. On the internet I've read the term "SSLeay format" a few times, and I think that might be how the format is called. The description in[1] seems to match.

PKCS#8 seems to encrypt the key first and then does the PEM encoding. So, after PEM decoding the data is binary, whether it was encrypted or not. In order to tell openssl whether it was encrypted or not the header is either "PRIVATE KEY" or "ENCRYPTED PRIVATE KEY".

Again, I am not 100% certain, but that is my understanding.

[1]: https://www.ibm.com/docs/en/secure-proxy/6.2.0?topic=certificates-certificate-formats-used-secure-proxy

Comment (weinimo) [reply pkg/octavia/amphora_certs.go:65]: I just found an example of a key that looks almost exactly as the `EncryptPEMBlock` generated key[1]. The only difference is it uses `RSA PRIVATE KEY`, which in turn is used for PKCS#1 too. It is confusing.

From now on I'll call the `EncryptPEMBlock` generated format PKCS#5.

[1]: https://stackoverflow.com/questions/52630701/how-to-encrypt-rsa-private-key-with-pbe-in-pkcs5-format-in-java-with-iaik-jce

File: pkg/octavia/pem_decrypt_fips.go (hunk 2 above)
Comment (weinimo) [reply pkg/octavia/pem_decrypt_fips.go]: No longer relevant. I am no longer copying code from the Golang stdlib.

**Comment by gthiemonge** [comment]
/lgtm

tested in non-FIPS env, no regression

**Pull request merged**